storage/storage_image.go

```diff
@@ -654,6 +667,14 @@ func (s *storageImageDestination) Commit(ctx context.Context, unparsedToplevel t
 		// this layer blob to its parent, if it has one, or the blob's hex value otherwise.
 		diffID, haveDiffID := s.blobDiffIDs[blob.Digest]
 		if !haveDiffID {
+			if lid, ok := s.blobLayerIDs[blob.Digest]; ok {
+				if layer, err := s.imageRef.transport.store.Layer(lid); layer != nil && err == nil {
+					// The exploded layer (chain) found in the store with the
+					// specified custom ID.
+					lastLayer = layer.ID
```

Comment:
This is wrong; a `*c/storage.Layer` object refers to the full chain (all parent layers + the layer identified by `diffID` on top).

But `blobLayerIDs` is built _only_ using the annotations, i.e. an image reference and a _single layer_ digest — the key of `blobLayerIDs` is just a _single layer_ digest; i.e. it can’t differentiate between chains (root1, sameDigest); (root2, …, sameDigest); and (sameDigest) with no parent.

That’s why the existing code always creates `lastLayer` with a deterministic ID that is computed from the full chain; ignoring that chain and setting `lastLayer` based on the top layer AFAICS just doesn’t work.

If there is some backing implementation that provides overlay-formatted single-layer data, c/storage will probably have to expose an operation that, given a parent `*c/storage.Layer` and some kind of ID of that single-layer data, creates a new `*c/storage.Layer` object that represents mounting of that overlay-formatted single-layer data on top of the submitted parent.

Reply:
> This is wrong; a *c/storage.Layer object refers to the full chain (all parent layers + the layer identified by diffID on top).

Fixed this.
Now `c/storage.Store` has a new API `PutFromAdditionalLayerStore(id, parent string, d digest.Digest, labels map[string]string) (*Layer, error)` for enabling `c/image` to always creates lastLayer with a ID that is computed from the full chain.